backend/assignments/1/src/main/java/CryptoMarket.java

```diff
@@ -0,0 +1,211 @@
+                    double price = Double.parseDouble(data[4].trim());
+                    long circulatingSupply = Long.parseLong(data[5].trim());
+
+                    Coin coin = new Coin(rank, name, symbol, price, circulatingSupply);
+                    coins.add(coin);
+                } catch (NumberFormatException e) {
+                    logger.info("Invalid rank value in CSV: " + rankStr);
+                }
+            } else {
+                logger.info("Rank value is empty in CSV.");
+            }
+        } else {
+            logger.info("Invalid data in CSV: ");
+        }
+    }
+    public void loadCoinsFromCSV(String filePath) {
+
+        coins = new ArrayList<>();
+
+
+        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
+            String line = br.readLine();
+
+            while (line != null) {
+                processCsvLine(line);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    public void loadTradersFromCSV(String filePath) {
+        traders = new ArrayList<>();
+
+
+        try(BufferedReader br = new BufferedReader(new FileReader(filePath)))
+        {
+            String line = br.readLine();
+
+            while (line != null) {
+                String[] data = line.split(",");
+                if (data.length >= 5) {
+                    String firstName = data[1].trim();
+                    String lastName = data[2].trim();
+                    String phone = data[3].trim();
+                    String walletAddress = data[4].trim();
+
+                    Trader trader = new Trader(firstName, lastName, phone, walletAddress);
+                    traders.add(trader);
+                }
+                else {
+                    logger.info("invalid data in CSV");
+                }
+            }
+        }catch(IOException e)
+        {
+            e.printStackTrace();
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    public synchronized void processBuyTransaction(JsonNode transactionNode) {
+        String coinCode = transactionNode.get("data").get("coin").asText();
+        int quantity = transactionNode.get("data").get("quantity").asInt();
+        String walletAddress = transactionNode.get("data").get("wallet_address").asText();
+
+        Coin coin = getCoinByCode(coinCode);
+        Trader trader = getTraderByWalletAddress(walletAddress);
+
+        if (coin != null && trader != null) {
+            if (coin.getVolume() >= quantity) {
+                coin.decreaseVolume(quantity);
+                trader.addToPortfolio(coinCode, quantity);
+                double totalCost = quantity * coin.getPrice();
+                trader.updateBalance(-totalCost);
+            } else {
+                logger.info("Insufficient volume for the BUY transaction. Transitioning to a pending state...");
+            }
+        }
+
+
+    }
```

Comment:
You should have implemented sync on coins / concurrentHashmaps.
Like this you are blocking other coins transactions as well